File: src/reference_code/world_model/model/multi_agents_trajectory_model.py
import sys

from sympy import *

sys.path.append("../../pnc_solver/scripts/")

from core.ocp_model import *

file_path = os.path.abspath(__file__)
dir_path = os.path.dirname(file_path)
solver_path = os.path.dirname(dir_path)
gen_dir = solver_path + "/model/multi_agents_trajectory_model/"
if len(sys.argv) > 1:
    gen_dir = sys.argv[1]

ocp_model = OptimalControlProblem("MultiAgentsTrajectoryModel")

epsilon = 1e-4

# define state symbols
# s 为投影到参考线上的匹配点的local s信息
ego_s, ego_x, ego_y, ego_theta, ego_v = ocp_model.add(
    "state", "ego_s, ego_x, ego_y, ego_theta, ego_v"
)
agent_0_s, agent_0_x, agent_0_y, agent_0_theta, agent_0_v = ocp_model.add(
    "state", "agent_0_s, agent_0_x, agent_0_y, agent_0_theta, agent_0_v"
)

# define control symbols
ego_a, ego_steer = ocp_model.add("ctrl", "ego_a, ego_steer")
agent_0_a = ocp_model.add("ctrl", "agent_0_a")


# define slack symbols
(
    ego_f_2_agent_0_slack,
    ego_r_2_agent_0_slack,
    ego_f_2_object_0_slack,
    ego_r_2_object_0_slack,
    ego_f_2_object_1_slack,
    ego_r_2_object_1_slack,
) = ocp_model.add(
    "slack",
    "ego_f_2_agent_0_slack, ego_r_2_agent_0_slack, ego_f_2_object_0_slack, ego_r_2_object_0_slack, ego_f_2_object_1_slack, ego_r_2_object_1_slack",
)


# reference line state params
ego_x_ref, ego_y_ref, ego_theta_ref, ego_kappa_ref, ego_steer_ref = ocp_model.add(
    "param", "ego_x_ref, ego_y_ref, ego_theta_ref, ego_kappa_ref, ego_steer_ref"
)
(
    agent_0_x_ref,
    agent_0_y_ref,
    agent_0_theta_ref,
    agent_0_kappa_ref,
    agent_0_steer_ref,
) = ocp_model.add(
    "param",
    "agent_0_x_ref, agent_0_y_ref, agent_0_theta_ref, agent_0_kappa_ref, agent_0_steer_ref",
)

#  coarse state params s_coarse 为决策轨迹在参考线上的投影点的local s信息
ego_s_coarse, ego_v_coarse, ego_a_coarse = ocp_model.add(
    "param", "ego_s_coarse, ego_v_coarse, ego_a_coarse"
)
agent_0_s_coarse, agent_0_v_coarse, agent_0_a_coarse = ocp_model.add(
    "param", "agent_0_s_coarse, agent_0_v_coarse, agent_0_a_coarse"
)

l_offset = ocp_model.add("param", l_offset=0.0)


wheelbase, front_overhang, rear_overhang, length, width = ocp_model.add(
    "param",
    wheelbase=3.0,
    front_overhang=1.0,
    rear_overhang=1.0,
    length=5.0,
    width=2.0,
)

# cost params

l_ref_weight, theta_ref_weight, steer_weight = ocp_model.add(
    "param",
    l_ref_weight=1.0,
    theta_ref_weight=1.0,
    steer_weight=1.0,
)


ego_s_coarse_weight, ego_v_coarse_weight, ego_a_weight = ocp_model.add(
    "param", ego_s_coarse_weight=1.0, ego_v_coarse_weight=1.0, ego_a_weight=1.0
)
agent_0_s_coarse_weight, agent_0_v_coarse_weight, agent_0_a_weight = ocp_model.add(
    "param",
    agent_0_s_coarse_weight=1.0,
    agent_0_v_coarse_weight=1.0,
    agent_0_a_weight=1.0,
)


agent_0_avoid_weight = ocp_model.add("param", agent_0_avoid_weight=100.0)

obs_0_weight, obs_1_weight = ocp_model.add("param", obs_0_weight=1.0, obs_1_weight=1.0)

Agent_0_AvoidedBound = ocp_model.add("param", Agent_0_AvoidedBound=1.0)

Obs_0_AvoidedBound, Obs_1_AvoidedBound = ocp_model.add(
    "param",
    Obs_0_AvoidedBound=1.0,
    Obs_1_AvoidedBound=1.0,
)

LHardLowerBound, LHardUpperBound = ocp_model.add(
    "param", LHardLowerBound=-20.0, LHardUpperBound=20.0
)


SteerLowerBound, SteerUpperBound = ocp_model.add(
    "param", SteerLowerBound=-0.436, SteerUpperBound=0.436
)  # 默认最大轮转角度为25度


SHardLowerBound, SHardUpperBound = ocp_model.add(
    "param", SHardLowerBound=0.0, SHardUpperBound=300.0
)

VHardLowerBound, VHardUpperBound = ocp_model.add(
    "param", VHardLowerBound=0.0, VHardUpperBound=40.0
)

AHardLowerBound, AHardUpperBound = ocp_model.add(
    "param", AHardLowerBound=-6.0, AHardUpperBound=3.0
)

Agent_0_AHardLowerBound, Agent_0_AHardUpperBound = ocp_model.add(
    "param", Agent_0_AHardLowerBound=-3.0, Agent_0_AHardUpperBound=3.0
)


# 计算点到当前障碍物椭圆的归一化距离
def compute_ellipse_distance(
    point_x,
    point_y,
    a_square,
    b_square,
    x_agent,
    y_agent,
    cos_theta_agent,
    sin_theta_agent,
):
    dx = point_x - x_agent
    dy = point_y - y_agent
    x_local = dx * cos_theta_agent + dy * sin_theta_agent
    y_local = -dx * sin_theta_agent + dy * cos_theta_agent
    return (x_local**2) / a_square + (y_local**2) / b_square


(
    agent_a_0,
    agent_a_0_square,
    agent_b_0,
    agent_b_0_square,
    agent_length_0,
    agent_width_0,
) = ocp_model.add(
    "param",
    agent_a_0=5.0,
    agent_a_0_square=25.0,
    agent_b_0=2.0,
    agent_b_0_square=4.0,
    agent_length_0=5.0,
    agent_width_0=2.0,
)


(
    object_a_0,
    object_a_0_square,
    object_b_0,
    object_b_0_square,
    object_x_0,
    object_y_0,
    object_theta_0,
    object_cos_theta_0,
    object_sin_theta_0,
    object_v_0,
    object_length_0,
    object_width_0,
) = ocp_model.add(
    "param",
    object_a_0=5.0,
    object_a_0_square=25.0,
    object_b_0=2.0,
    object_b_0_square=4.0,
    object_x_0=0.0,
    object_y_0=0.0,
    object_theta_0=0.0,
    object_cos_theta_0=1.0,
    object_sin_theta_0=0.0,
    object_v_0=0.0,
    object_length_0=5.0,
    object_width_0=2.0,
)

(
    object_a_1,
    object_a_1_square,
    object_b_1,
    object_b_1_square,
    object_x_1,
    object_y_1,
    object_theta_1,
    object_cos_theta_1,
    object_sin_theta_1,
    object_v_1,
    object_length_1,
    object_width_1,
) = ocp_model.add(
    "param",
    object_a_1=5.0,
    object_a_1_square=25.0,
    object_b_1=2.0,
    object_b_1_square=4.0,
    object_x_1=0.0,
    object_y_1=0.0,
    object_theta_1=0.0,
    object_cos_theta_1=1.0,
    object_sin_theta_1=0.0,
    object_v_1=0.0,
    object_length_1=5.0,
    object_width_1=2.0,
)


# define dynamics
# for ego
ego_l = -sin(ego_theta_ref) * (ego_x - ego_x_ref) + cos(ego_theta_ref) * (
    ego_y - ego_y_ref
)
ego_ds = cos(ego_theta - ego_theta_ref) / (1 - ego_kappa_ref * ego_l)
ego_kappa = tan(ego_steer) / wheelbase

ocp_model.subject(dot(ego_s) == ego_v * ego_ds)
ocp_model.subject(dot(ego_x) == ego_v * cos(ego_theta))
ocp_model.subject(dot(ego_y) == ego_v * sin(ego_theta))
ocp_model.subject(dot(ego_theta) == ego_v * ego_kappa)
ocp_model.subject(dot(ego_v) == ego_a)


# for agent 0
ocp_model.subject(dot(agent_0_s) == agent_0_v)
ocp_model.subject(dot(agent_0_x) == agent_0_v * cos(agent_0_theta_ref))
ocp_model.subject(dot(agent_0_y) == agent_0_v * sin(agent_0_theta_ref))
ocp_model.subject(dot(agent_0_v) == agent_0_a)


# 车辆头部中心坐标
x_f_circle = ego_x + (wheelbase + front_overhang) * cos(ego_theta)

y_f_circle = ego_y + (wheelbase + front_overhang) * sin(ego_theta)

# 车辆尾部中心坐标
x_r_circle = ego_x - rear_overhang * cos(ego_theta)

y_r_circle = ego_y - rear_overhang * sin(ego_theta)


# for agent
f_dis_2_agent_0 = compute_ellipse_distance(
    x_f_circle,
    y_f_circle,
    agent_a_0_square,
    agent_b_0_square,
    agent_0_x,
    agent_0_y,
    cos(agent_0_theta_ref),
    sin(agent_0_theta_ref),
)
r_dis_2_agent_0 = compute_ellipse_distance(
    x_r_circle,
    y_r_circle,
    agent_a_0_square,
    agent_b_0_square,
    agent_0_x,
    agent_0_y,
    cos(agent_0_theta_ref),
    sin(agent_0_theta_ref),
)


# for object
f_dis_2_object_0 = compute_ellipse_distance(
    x_f_circle,
    y_f_circle,
    object_a_0_square,
    object_b_0_square,
    object_x_0,
    object_y_0,
    object_cos_theta_0,
    object_sin_theta_0,
)
r_dis_2_object_0 = compute_ellipse_distance(
    x_r_circle,
    y_r_circle,
    object_a_0_square,
    object_b_0_square,
    object_x_0,
    object_y_0,
    object_cos_theta_0,
    object_sin_theta_0,
)

f_dis_2_object_1 = compute_ellipse_distance(
    x_f_circle,
    y_f_circle,
    object_a_1_square,
    object_b_1_square,
    object_x_1,
    object_y_1,
    object_cos_theta_1,
    object_sin_theta_1,
)
r_dis_2_object_1 = compute_ellipse_distance(
    x_r_circle,
    y_r_circle,
    object_a_1_square,
    object_b_1_square,
    object_x_1,
    object_y_1,
    object_cos_theta_1,
    object_sin_theta_1,
)


(
    s_scale,
    v_scale,
    a_scale,
    jerk_scale,
    l_scale,
    theta_scale,
    steer_scale,
) = ocp_model.add(
    "param",
    s_scale=20.0,
    v_scale=10.0,
    a_scale=5.0,
    jerk_scale=5.0,
    l_scale=1.0,
    theta_scale=1.0,
    steer_scale=1.0,
)

# define cost
cost_x_lateral = (
    0.5 * l_ref_weight * ((ego_l - l_offset) / l_scale) ** 2
    + 0.5 * theta_ref_weight * ((ego_theta - ego_theta_ref) / theta_scale) ** 2
)

cost_x_longitudinal = (
    +0.5 * ego_v_coarse_weight * ((ego_v - ego_v_coarse) / v_scale) ** 2
    + 0.5 * agent_0_v_coarse_weight * ((agent_0_v - agent_0_v_coarse) / v_scale) ** 2
    + 0.5 * ego_s_coarse_weight * ((ego_s - ego_s_coarse) / s_scale) ** 2
    + 0.5 * agent_0_s_coarse_weight * ((agent_0_s - agent_0_s_coarse) / s_scale) ** 2
)

cost_u = (
    0.5 * steer_weight * ((ego_steer - ego_steer_ref) / steer_scale) ** 2
    + 0.5 * ego_a_weight * (ego_a / a_scale) ** 2
    + 0.5 * agent_0_a_weight * (agent_0_a / a_scale) ** 2
)

cost_s = (
    0.5 * agent_0_avoid_weight * ego_f_2_agent_0_slack
    + 0.5 * agent_0_avoid_weight * ego_r_2_agent_0_slack
    + 0.5 * obs_0_weight * ego_f_2_object_0_slack
    + 0.5 * obs_0_weight * ego_r_2_object_0_slack
    + 0.5 * obs_1_weight * ego_f_2_object_1_slack
    + 0.5 * obs_1_weight * ego_r_2_object_1_slack
)

ocp_model.add("cost", cost_x_lateral, cost_x_longitudinal, cost_u, cost_s)


# define constraints
# for agent
ocp_model.subject(Agent_0_AvoidedBound <= f_dis_2_agent_0 + ego_f_2_agent_0_slack)


ocp_model.subject(Agent_0_AvoidedBound <= r_dis_2_agent_0 + ego_r_2_agent_0_slack)


# for object
ocp_model.subject(Obs_0_AvoidedBound <= f_dis_2_object_0 + ego_f_2_object_0_slack)

ocp_model.subject(Obs_1_AvoidedBound <= f_dis_2_object_1 + ego_f_2_object_1_slack)

ocp_model.subject(Obs_0_AvoidedBound <= r_dis_2_object_0 + ego_r_2_object_0_slack)

ocp_model.subject(Obs_1_AvoidedBound <= r_dis_2_object_1 + ego_r_2_object_1_slack)

# for slack
ocp_model.subject(ego_f_2_agent_0_slack >= 0.0)
ocp_model.subject(ego_r_2_agent_0_slack >= 0.0)
ocp_model.subject(ego_f_2_object_0_slack >= 0.0)
ocp_model.subject(ego_r_2_object_0_slack >= 0.0)
ocp_model.subject(ego_f_2_object_1_slack >= 0.0)
ocp_model.subject(ego_r_2_object_1_slack >= 0.0)

ocp_model.subject(LHardLowerBound <= ego_l, ego_l <= LHardUpperBound)


ocp_model.subject(SteerLowerBound <= ego_steer, ego_steer <= SteerUpperBound)


ocp_model.subject(SHardLowerBound <= ego_s, ego_s <= SHardUpperBound)
ocp_model.subject(agent_0_s >= 0.0)

ocp_model.subject(VHardLowerBound <= ego_v, ego_v <= VHardUpperBound)
ocp_model.subject(agent_0_v >= 0.0)

ocp_model.subject(AHardLowerBound <= ego_a, ego_a <= AHardUpperBound)
ocp_model.subject(
    Agent_0_AHardLowerBound <= agent_0_a, agent_0_a <= Agent_0_AHardUpperBound
)

ocp_model.subject(1 - ego_kappa_ref * ego_l >= epsilon)
ocp_model.subject(cos(ego_theta - ego_theta_ref) >= 0.0)


ocp_model.generate(gen_dir)

File: src/reference_code/world_model/model/multi_agents_trajectory_model/multi_agents_trajectory_model.cpp
#include "multi_agents_trajectory_model.h"

namespace gpal::pnc::planning {

MultiAgentsTrajectoryModel::MultiAgentsTrajectoryModel() : OptimalControlProblem("MultiAgentsTrajectoryModel", 10, 9, 6, 86, 13) {
  default_state_.emplace("ego_s");
  default_state_.emplace("ego_x");
  default_state_.emplace("ego_y");
  default_state_.emplace("ego_theta");
  default_state_.emplace("ego_v");
  default_state_.emplace("agent_0_s");
  default_state_.emplace("agent_0_x");
  default_state_.emplace("agent_0_y");
  default_state_.emplace("agent_0_theta");
  default_state_.emplace("agent_0_v");

  default_ctrl_.emplace("ego_a");
  default_ctrl_.emplace("ego_steer");
  default_ctrl_.emplace("agent_0_a");
  default_ctrl_.emplace("ego_f_2_agent_0_slack");
  default_ctrl_.emplace("ego_r_2_agent_0_slack");
  default_ctrl_.emplace("ego_f_2_object_0_slack");
  default_ctrl_.emplace("ego_r_2_object_0_slack");
  default_ctrl_.emplace("ego_f_2_object_1_slack");
  default_ctrl_.emplace("ego_r_2_object_1_slack");

  default_param_.emplace("ego_x_ref", 0.0);
  default_param_.emplace("ego_y_ref", 0.0);
  default_param_.emplace("ego_theta_ref", 0.0);
  default_param_.emplace("ego_kappa_ref", 0.0);
  default_param_.emplace("ego_steer_ref", 0.0);
  default_param_.emplace("agent_0_x_ref", 0.0);
  default_param_.emplace("agent_0_y_ref", 0.0);
  default_param_.emplace("agent_0_theta_ref", 0.0);
  default_param_.emplace("agent_0_kappa_ref", 0.0);
  default_param_.emplace("agent_0_steer_ref", 0.0);
  default_param_.emplace("ego_s_coarse", 0.0);
  default_param_.emplace("ego_v_coarse", 0.0);
  default_param_.emplace("ego_a_coarse", 0.0);
  default_param_.emplace("agent_0_s_coarse", 0.0);
  default_param_.emplace("agent_0_v_coarse", 0.0);
  default_param_.emplace("agent_0_a_coarse", 0.0);
  default_param_.emplace("l_offset", 0.0);
  default_param_.emplace("wheelbase", 3.0);
  default_param_.emplace("front_overhang", 1.0);
  default_param_.emplace("rear_overhang", 1.0);
  default_param_.emplace("length", 5.0);
  default_param_.emplace("width", 2.0);
  default_param_.emplace("l_ref_weight", 1.0);
  default_param_.emplace("theta_ref_weight", 1.0);
  default_param_.emplace("steer_weight", 1.0);
  default_param_.emplace("ego_s_coarse_weight", 1.0);
  default_param_.emplace("ego_v_coarse_weight", 1.0);
  default_param_.emplace("ego_a_weight", 1.0);
  default_param_.emplace("agent_0_s_coarse_weight", 1.0);
  default_param_.emplace("agent_0_v_coarse_weight", 1.0);
  default_param_.emplace("agent_0_a_weight", 1.0);
  default_param_.emplace("agent_0_avoid_weight", 100.0);
  default_param_.emplace("obs_0_weight", 1.0);
  default_param_.emplace("obs_1_weight", 1.0);
  default_param_.emplace("Agent_0_AvoidedBound", 1.0);
  default_param_.emplace("Obs_0_AvoidedBound", 1.0);
  default_param_.emplace("Obs_1_AvoidedBound", 1.0);
  default_param_.emplace("LHardLowerBound", -20.0);
  default_param_.emplace("LHardUpperBound", 20.0);
  default_param_.emplace("SteerLowerBound", -0.436);
  default_param_.emplace("SteerUpperBound", 0.436);
  default_param_.emplace("SHardLowerBound", 0.0);
  default_param_.emplace("SHardUpperBound", 300.0);
  default_param_.emplace("VHardLowerBound", 0.0);
  default_param_.emplace("VHardUpperBound", 40.0);
  default_param_.emplace("AHardLowerBound", -6.0);
  default_param_.emplace("AHardUpperBound", 3.0);
  default_param_.emplace("Agent_0_AHardLowerBound", -3.0);
  default_param_.emplace("Agent_0_AHardUpperBound", 3.0);
  default_param_.emplace("agent_a_0", 5.0);
  default_param_.emplace("agent_a_0_square", 25.0);
  default_param_.emplace("agent_b_0", 2.0);
  default_param_.emplace("agent_b_0_square", 4.0);
  default_param_.emplace("agent_length_0", 5.0);
  default_param_.emplace("agent_width_0", 2.0);
  default_param_.emplace("object_a_0", 5.0);
  default_param_.emplace("object_a_0_square", 25.0);
  default_param_.emplace("object_b_0", 2.0);
  default_param_.emplace("object_b_0_square", 4.0);
  default_param_.emplace("object_x_0", 0.0);
  default_param_.emplace("object_y_0", 0.0);
  default_param_.emplace("object_theta_0", 0.0);
  default_param_.emplace("object_cos_theta_0", 1.0);
  default_param_.emplace("object_sin_theta_0", 0.0);
  default_param_.emplace("object_v_0", 0.0);
  default_param_.emplace("object_length_0", 5.0);
  default_param_.emplace("object_width_0", 2.0);
  default_param_.emplace("object_a_1", 5.0);
  default_param_.emplace("object_a_1_square", 25.0);
  default_param_.emplace("object_b_1", 2.0);
  default_param_.emplace("object_b_1_square", 4.0);
  default_param_.emplace("object_x_1", 0.0);
  default_param_.emplace("object_y_1", 0.0);
  default_param_.emplace("object_theta_1", 0.0);
  default_param_.emplace("object_cos_theta_1", 1.0);
  default_param_.emplace("object_sin_theta_1", 0.0);
  default_param_.emplace("object_v_1", 0.0);
  default_param_.emplace("object_length_1", 5.0);
  default_param_.emplace("object_width_1", 2.0);
  default_param_.emplace("s_scale", 20.0);
  default_param_.emplace("v_scale", 10.0);
  default_param_.emplace("a_scale", 5.0);
  default_param_.emplace("jerk_scale", 5.0);
  default_param_.emplace("l_scale", 1.0);
  default_param_.emplace("theta_scale", 1.0);
  default_param_.emplace("steer_scale", 1.0);

  default_global_.emplace("ego_kappa");
  default_global_.emplace("x_r_circle");
  default_global_.emplace("y_r_circle");
  default_global_.emplace("ego_l");
  default_global_.emplace("x_f_circle");
  default_global_.emplace("y_f_circle");
  default_global_.emplace("ego_ds");
  default_global_.emplace("r_dis_2_agent_0");
  default_global_.emplace("r_dis_2_object_0");
  default_global_.emplace("r_dis_2_object_1");
  default_global_.emplace("f_dis_2_agent_0");
  default_global_.emplace("f_dis_2_object_0");
  default_global_.emplace("f_dis_2_object_1");
  default_global_.setUpdater(std::make_shared<OcpVariable::Updater>([](const OcpStage& stage, OcpVariable* ptr_global) {
    // Determine stage variables
    const auto& ego_x = stage.x(1);
    const auto& ego_y = stage.x(2);
    const auto& ego_theta = stage.x(3);
    const auto& agent_0_x = stage.x(6);
    const auto& agent_0_y = stage.x(7);
    const auto& ego_steer = stage.u(1);
    const auto& ego_x_ref = stage.params(0);
    const auto& ego_y_ref = stage.params(1);
    const auto& ego_theta_ref = stage.params(2);
    const auto& ego_kappa_ref = stage.params(3);
    const auto& agent_0_theta_ref = stage.params(7);
    const auto& wheelbase = stage.params(17);
    const auto& front_overhang = stage.params(18);
    const auto& rear_overhang = stage.params(19);
    const auto& agent_a_0_square = stage.params(50);
    const auto& agent_b_0_square = stage.params(52);
    const auto& object_a_0_square = stage.params(56);
    const auto& object_b_0_square = stage.params(58);
    const auto& object_x_0 = stage.params(59);
    const auto& object_y_0 = stage.params(60);
    const auto& object_cos_theta_0 = stage.params(62);
    const auto& object_sin_theta_0 = stage.params(63);
    const auto& object_a_1_square = stage.params(68);
    const auto& object_b_1_square = stage.params(70);
    const auto& object_x_1 = stage.params(71);
    const auto& object_y_1 = stage.params(72);
    const auto& object_cos_theta_1 = stage.params(74);
    const auto& object_sin_theta_1 = stage.params(75);

    (*ptr_global)(0) = tan(ego_steer)/wheelbase;
    (*ptr_global)(1) = ego_x - rear_overhang*cos(ego_theta);
    (*ptr_global)(2) = ego_y - rear_overhang*sin(ego_theta);
    (*ptr_global)(3) = -(ego_x - ego_x_ref)*sin(ego_theta_ref) + (ego_y - ego_y_ref)*cos(ego_theta_ref);
    (*ptr_global)(4) = ego_x + (front_overhang + wheelbase)*cos(ego_theta);
    (*ptr_global)(5) = ego_y + (front_overhang + wheelbase)*sin(ego_theta);
    (*ptr_global)(6) = cos(ego_theta - ego_theta_ref)/(-ego_kappa_ref*(-(ego_x - ego_x_ref)*sin(ego_theta_ref) + (ego_y - ego_y_ref)*cos(ego_theta_ref)) + 1);
    (*ptr_global)(7) = pow((agent_0_x - ego_x + rear_overhang*cos(ego_theta))*sin(agent_0_theta_ref) + (-agent_0_y + ego_y - rear_overhang*sin(ego_theta))*cos(agent_0_theta_ref), 2)/agent_b_0_square + pow((-agent_0_x + ego_x - rear_overhang*cos(ego_theta))*cos(agent_0_theta_ref) + (-agent_0_y + ego_y - rear_overhang*sin(ego_theta))*sin(agent_0_theta_ref), 2)/agent_a_0_square;
    (*ptr_global)(8) = pow(object_cos_theta_0*(ego_y - object_y_0 - rear_overhang*sin(ego_theta)) + object_sin_theta_0*(-ego_x + object_x_0 + rear_overhang*cos(ego_theta)), 2)/object_b_0_square + pow(object_cos_theta_0*(ego_x - object_x_0 - rear_overhang*cos(ego_theta)) + object_sin_theta_0*(ego_y - object_y_0 - rear_overhang*sin(ego_theta)), 2)/object_a_0_square;
    (*ptr_global)(9) = pow(object_cos_theta_1*(ego_y - object_y_1 - rear_overhang*sin(ego_theta)) + object_sin_theta_1*(-ego_x + object_x_1 + rear_overhang*cos(ego_theta)), 2)/object_b_1_square + pow(object_cos_theta_1*(ego_x - object_x_1 - rear_overhang*cos(ego_theta)) + object_sin_theta_1*(ego_y - object_y_1 - rear_overhang*sin(ego_theta)), 2)/object_a_1_square;
    (*ptr_global)(10) = pow((agent_0_x - ego_x - (front_overhang + wheelbase)*cos(ego_theta))*sin(agent_0_theta_ref) + (-agent_0_y + ego_y + (front_overhang + wheelbase)*sin(ego_theta))*cos(agent_0_theta_ref), 2)/agent_b_0_square + pow((-agent_0_x + ego_x + (front_overhang + wheelbase)*cos(ego_theta))*cos(agent_0_theta_ref) + (-agent_0_y + ego_y + (front_overhang + wheelbase)*sin(ego_theta))*sin(agent_0_theta_ref), 2)/agent_a_0_square;
    (*ptr_global)(11) = pow(object_cos_theta_0*(ego_y - object_y_0 + (front_overhang + wheelbase)*sin(ego_theta)) + object_sin_theta_0*(-ego_x + object_x_0 - (front_overhang + wheelbase)*cos(ego_theta)), 2)/object_b_0_square + pow(object_cos_theta_0*(ego_x - object_x_0 + (front_overhang + wheelbase)*cos(ego_theta)) + object_sin_theta_0*(ego_y - object_y_0 + (front_overhang + wheelbase)*sin(ego_theta)), 2)/object_a_0_square;
    (*ptr_global)(12) = pow(object_cos_theta_1*(ego_y - object_y_1 + (front_overhang + wheelbase)*sin(ego_theta)) + object_sin_theta_1*(-ego_x + object_x_1 - (front_overhang + wheelbase)*cos(ego_theta)), 2)/object_b_1_square + pow(object_cos_theta_1*(ego_x - object_x_1 + (front_overhang + wheelbase)*cos(ego_theta)) + object_sin_theta_1*(ego_y - object_y_1 + (front_overhang + wheelbase)*sin(ego_theta)), 2)/object_a_1_square;
  }));
}

std::shared_ptr<Dynamics> MultiAgentsTrajectoryModel::createDynamics(OcpConfig::IntegratorType type) const {
  return std::make_shared<MultiAgentsTrajectoryModelDynamics>(type);
}

std::shared_ptr<CostFunction> MultiAgentsTrajectoryModel::createCostFunction() const {
  return std::make_shared<MultiAgentsTrajectoryModelCost<StageType::NORMINAL>>();
}

std::shared_ptr<Constraints> MultiAgentsTrajectoryModel::createConstraint() const {
  return std::make_shared<MultiAgentsTrajectoryModelConstraints<StageType::NORMINAL>>();
}

std::shared_ptr<Constraints> MultiAgentsTrajectoryModel::createInitialConstraint() const {
  return std::make_shared<MultiAgentsTrajectoryModelConstraints<StageType::INITIAL>>();
}

std::shared_ptr<Constraints> MultiAgentsTrajectoryModel::createTerminalConstraint() const {
  return std::make_shared<MultiAgentsTrajectoryModelConstraints<StageType::TERMINAL>>();
}

std::shared_ptr<Constraints> MultiAgentsTrajectoryModel::createStateOnlyEqualities() const {
  return std::make_shared<MultiAgentsTrajectoryModelStateOnlyEqualities<StageType::NORMINAL>>();
}

std::shared_ptr<Constraints> MultiAgentsTrajectoryModel::createInitialStateOnlyEqualities() const {
  return std::make_shared<MultiAgentsTrajectoryModelStateOnlyEqualities<StageType::INITIAL>>();
}

std::shared_ptr<Constraints> MultiAgentsTrajectoryModel::createTerminalEqualities() const {
  return std::make_shared<MultiAgentsTrajectoryModelStateOnlyEqualities<StageType::TERMINAL>>();
}

std::shared_ptr<IPMHelper> MultiAgentsTrajectoryModel::createBaseIpmHelper() const {
  IPMHelper::Creater creater{
    {{StageType::NORMINAL, OcpConfig::FORWARD_EULER},[]() { return std::make_shared<MultiAgentsTrajectoryModelIpmEvaluator<StageType::NORMINAL, OcpConfig::FORWARD_EULER>>(); }},
    {{StageType::NORMINAL, OcpConfig::ERK4},[]() { return std::make_shared<MultiAgentsTrajectoryModelIpmEvaluator<StageType::NORMINAL, OcpConfig::ERK4>>(); }},
    {{StageType::INITIAL, OcpConfig::FORWARD_EULER},[]() { return std::make_shared<MultiAgentsTrajectoryModelIpmEvaluator<StageType::INITIAL, OcpConfig::FORWARD_EULER>>(); }},
    {{StageType::INITIAL, OcpConfig::ERK4},[]() { return std::make_shared<MultiAgentsTrajectoryModelIpmEvaluator<StageType::INITIAL, OcpConfig::ERK4>>(); }},
    {{StageType::TERMINAL, OcpConfig::FORWARD_EULER},[]() { return std::make_shared<MultiAgentsTrajectoryModelIpmEvaluatorTerminal>(); }},
    {{StageType::TERMINAL, OcpConfig::ERK4},[]() { return std::make_shared<MultiAgentsTrajectoryModelIpmEvaluatorTerminal>(); }}};
  return std::make_shared<IPMHelper>(std::move(creater));
}

REGIST_MODEL(MultiAgentsTrajectoryModel)
} // namespace gpal::pnc::planning